vcd/resource_vcd_vapp_nat_rules.go

```diff
@@ -0,0 +1,310 @@
+package vcd
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"log"
+)
+
+const (
+	allowTrafficInPolicy  = "allowTrafficIn"
+	allowTrafficPolicy    = "allowTraffic"
+	ipTranslationNatType  = "ipTranslation"
+	portForwardingNatType = "portForwarding"
+)
+
+func resourceVcdVappNetworkNatRules() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVappNetworkNatRulesCreate,
+		Delete: resourceVAppNetworkNatRulesDelete,
+		Read:   resourceVappNetworkNatRulesRead,
+		Update: resourceVappNetworkNatRulesUpdate,
+		Importer: &schema.ResourceImporter{
+			State: vappNetworkNatRuleImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"org": {
+				Type:     schema.TypeString,
+				Required: false,
+				Optional: true,
+				ForceNew: true,
+				Description: "The name of organization to use, optional if defined at provider " +
+					"level. Useful when connected as sysadmin working across different organizations",
+			},
+			"vdc": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			"vapp_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "vApp identifier",
+			},
+			"network_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "vApp network identifier",
+			},
+			"nat_type": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{portForwardingNatType, ipTranslationNatType}, false),
+				Description:  "One of: `ipTranslation` (use IP translation), `portForwarding` (use port forwarding).",
+			},
+			"enable_ip_masquerade": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "When enabled translates a virtual machine's private, internal IP address to a public IP address for outbound traffic.",
+			},
+			"rule": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Id of the rule. Can be used to track syslog messages.",
+						},
+						"mapping_mode": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice([]string{"automatic", "manual"}, false),
+							Description:  "Mapping mode. One of: `automatic`, `manual`",
+						},
+						"vm_id": &schema.Schema{
+							Type:        schema.TypeString,
+							Optional:    true,
```

Comment:
I tried to create a rule without `vm_id`:
```
resource "vcd_vapp_nat_rules" "vapp_nat" {
  vapp_id              = vcd_vapp.demo_vapp.id
  network_id           = vcd_vapp_network.demo_vapp_net_w_org.id
  nat_type             = "portForwarding"
  enable_ip_masquerade = false

  rule {
    external_port   = "22"
    forward_to_port = "-1"
    protocol        = "TCP_UDP"
    vm_nic_id       = "0"
    vm_id           = vcd_vapp_vm.demo_vm_wordpress.id
  }

  rule {
    external_port   = "-1"
    forward_to_port = "80"
    protocol        = "TCP"
    vm_nic_id       = "0"
    #vm_id           = vcd_vapp_vm.demo_vm_wordpress.id
  }
}
```
However, it didn't work:
```
vcd_vapp_nat_rules.vapp_nat: Creating...

Error: error expanding NAT rules: error fetchining VM: [ENF] entity not found

  on radio2020.tf line 292, in resource "vcd_vapp_nat_rules" "vapp_nat":
 292: resource "vcd_vapp_nat_rules" "vapp_nat" {
```

Reply:
I updated documentation what are needed fields by type, but I change also in Schema that vm id and nic id is mandatory.